### ISSUE TITLE
Connection Leaks in 4.7.1

### ISSUE THREAD

**Issue by tsf1** [MarimerLLC/cslaforum#518]
I updated to release 4.7.1 from 4.6.x for a MVC site in hopes of resolving database connection leak issues (refer to issue #377). We're still seeing the same problems after this update. I'm wondering if I have incorrectly coded my use of the ConnectionManager. The following code snippets show how we're using CSLA in a n-tier environment (Browser <> DMZ IIS Server <> Internal IIS Server <> SQL Database).

Here's our DalManager:
```
    public class DalManager : DataAccess.IDalManager
    {
        private static string _typeMask = typeof(DalManager).FullName.Replace("DalManager", @"{0}");

        public T GetProvider<T>() where T : class
        {
            var typeName = string.Format(_typeMask, typeof(T).Name.Substring(1));
            var type = Type.GetType(typeName);
            if (type != null)
                return Activator.CreateInstance(type) as T;
            else
                throw new NotImplementedException(typeName);
        }

        public ConnectionManager<SqlConnection> ConnectionManager { get; private set; }

        public DalManager()
        {
            ConnectionManager = ConnectionManager<SqlConnection>.GetManager("ConnectionString");
        }

        public void Dispose()
        {
            if (ConnectionManager != null)
            {
                ConnectionManager.Dispose();
                ConnectionManager = null;
            }
        }
    }
```

And here's the DalFactory:

```
    public static class DalFactory
    {
        private static Type _dalType;
        public static IDalManager GetManager()
        {
            bool reloadDal = false;

            if ((_dalType == null) || (reloadDal))
            {
                string dalTypeName = "Customer.DataAccess.Sql.DalManager,Customer.DataAccess.Sql";

                if (!string.IsNullOrEmpty(dalTypeName))
                    _dalType = Type.GetType(dalTypeName);
                else
                    throw new NullReferenceException("DalManagerType");
                if (_dalType == null)
                    throw new ArgumentException(string.Format("Type {0} could not be found", dalTypeName));
            }

            return (IDalManager)Activator.CreateInstance(_dalType);
        }
    }
```

Here's the typical flow for DP methods that get executed on our internal IIS site:
```
        private void DataPortal_Fetch(UserCriteria criteria)
        {
            using (var dalManager = DataAccess.DalFactory.GetManager())
            {
                var dal = dalManager.GetProvider<DataAccess.IAccountDal>();

                try
                {
                    using (var data = new Csla.Data.SafeDataReader(dal.GetAccount(criteria.AccountId)))
                    {
                        if (!data.IsClosed && data.Read())
                        {
                            UserName = data.GetString(data.GetOrdinal("Name"));
                            EmailAddress = data.GetString(data.GetOrdinal("EmailAddress"));
                        }
                        data.Close();
                    }
                }
                catch (Exception ex)
                {
                    // log exception info and throw;
                }

                // get the list of stores associated to this user account
                AssociatedStoreList = CustomerInfoList.GetAssociatedCustomers(criteria.AccountId);
            }            
        }
```

And finally, here's the typical DAL method when calling stored procs:
```
        public IDataReader GetAccount(int accountId)
        {
            using (var ctx = ConnectionManager<SqlConnection>.GetManager("ConnectionString"))
            {
                var cm = ctx.Connection.CreateCommand();
                cm.CommandType = CommandType.StoredProcedure;
                cm.Parameters.AddWithValue("@AccountId", accountId);
                cm.CommandText = "customer.getAccount";

                return cm.ExecuteReader();
            }
        }
```

**Does anything look wrong with that code above? Am I missing something?**

Our evidence of leaked connections is that we are encountering the following errors throughout the day: "Timeout expired. The timeout period elapsed prior to obtaining a connection from the pool. This may have occurred because all pooled connections were in use and max pool size was reached."

When we run the following script in SQL we can easily see the number of connections going up and up until problems occur, even when only a few users are on the site. Additionally, some of our BO DP methods make a number of calls to other BOs (and their own database calls), and when these are hit you usually see a big jump in open SQL connections.
```
select spid, blocked, waittime, cpu, login_time, last_batch
		,status, program_name, hostprocess, cmd
from sysprocesses
where program_name = 'MyWebApp'
```

Because of the above, we have set the recycle interval for our app pool on our internal IIS server to a low setting, but that is a band aid that is only partially helping.

Thanks for your help.
Tim

**Comment by rockfordlhotka** [comment]
The only thing that looks suspicious to me is this:

```c#
        public DalManager()
        {
            ConnectionManager = ConnectionManager<SqlConnection>.GetManager("ConnectionString");
        }
```

iirc, each call to `GetManager` increases the reference count by one. And each call to `ConnectionManager.Dispose()` decreases it by one. You call `Dispose` exactly once, but this code offers the chance for `GetManager` to be called many times I think?

**Comment by tsf1** [comment]
What should it looks like? Should that not even be in there? I had that due to what your Data Access e-book included on pg 93:

![image](https://user-images.githubusercontent.com/2549019/38582252-2a7e7ac4-3cdd-11e8-8a32-9c59e753d3d7.png)

I just realized it was referring to SqlCe, but didn't think that would make a difference?

**Comment by rockfordlhotka** [comment]
Sorry - was multitasking while commenting - not the best plan - I think this _is_ correct, because each instance of `DalManager` gets exactly one `ConnectionManager`, so there should be a 1:1 relationship, and each `Dispose` of a manager should deref the connection it contains.

Sorry for the confusion...

**Comment by rockfordlhotka** [comment]
Though this is closed/resolved, is it possible the issue is related to this?

https://github.com/MarimerLLC/csla/issues/377

**Comment by tsf1** [comment]
Yes, I believe so. (I had mentioned that at the beginning of my first post - #377). But that thread seems to be related to how to change the underlying ConnectionManager class within the CSLA framework itself and not a work-around for those consuming it? Or at least it seems that way when I read it.

I was starting to consider changing my code to explicitly open/close connections and forgo the convenience of the ConnectionManager class, but if there is something I can do to work within the bounds of that, I'd prefer to.

**Comment by tsf1** [comment]
Sorry, just realized my use of "#377" was actually linking to something else...not sure how that happened. But 377 is the issue number, I believe.

**Comment by rockfordlhotka** [comment]
Different repos - this is the cslaforum repo, the 377 you want is in the csla repo.

**Comment by tsf1** [comment]
Ah. Either way, it seems that it was referring to changes you and johnnybee were considering to make inside CSLA. I re-read the post again and trying to understand what I can take from it to adjust on my end. Or - would it be better to abandon ConnectionManager and try using SqlConnection directly?

**Comment by tsf1** [comment]
I've read on various threads ([particularly this older one from William Vaughn](https://social.msdn.microsoft.com/Forums/en-US/c57c0432-c27b-45ab-81ca-b2df76c911ef/timeout-expired-the-timeout-period-elapsed-prior-to-obtaining-a-connection-from-the-pool?forum=adodotnetdataproviders)) that the Usings statement isn't always certain to properly close & dispose of connections. Not sure if that is still legit today, but it makes me wonder if I need to go that route with my code to see if anything changes. It is a lot of effort but our current status is not tenable.

**Comment by rockfordlhotka** [comment]
I've reopened 377 and put some thoughts there. This threading stuff is tricky!
